src/contexts/addresses.tsx

```diff
@@ -93,26 +98,55 @@ export class Address {
   }
 
   async fetchDetails(client: Client) {
-    if (!client) return
+    if (!client) throw new Error('No client provided')
```

Comment:
We pass `client` to each methods and return an error if none are provided. Do you see a way to make it dryer?

Reply:
Also, the client depends solely on the `network`, which is stored in the client class. 
Please see the comment below about the `custom` network that could cause issues.

Reply:
I think I found an elegant solution to this. Instead of passing `Client` as an argument to `Address` functions, I think it makes much more semantic sense to pass `Address` to `Client` functions. See the following diff to understand what I mean. I feel this is a much cleaner approach, keeping concerns of storing address state and querying the API separately. Let me know your thoughts.

<details><summary>Click here plz</summary>

```diff
diff --git a/src/contexts/addresses.tsx b/src/contexts/addresses.tsx
index 74ffd53..7ac5237 100644
--- a/src/contexts/addresses.tsx
+++ b/src/contexts/addresses.tsx
@@ -32,7 +32,7 @@ import { getHumanReadableError } from '../utils/api'
 import { NetworkType } from '../utils/settings'
 import { Client, useGlobalContext } from './global'
 
-type TransactionType = 'consolidation' | 'transfer' | 'sweep'
+export type TransactionType = 'consolidation' | 'transfer' | 'sweep'
 
 type SimpleTx = {
   txId: string
@@ -96,58 +96,6 @@ export class Address {
     return `${this.settings.isMain ? '★ ' : ''}${this.displayName()}`
   }
 
-  async fetchDetails(client: Client) {
-    if (!client) throw new Error('No client provided')
-    console.log('⬇️ Fetching address details: ', this.hash)
-
-    const { data } = await client.explorer.getAddressDetails(this.hash)
-    this.details = data
-
-    if (data.balance) this.availableBalance = BigInt(data.balance)
-    if (data.lockedBalance) this.availableBalance -= BigInt(data.lockedBalance)
-
-    return data
-  }
-
-  async fetchConfirmedTransactions(client: Client, page = 1) {
-    if (!client) throw new Error('No client provided')
-    console.log(`⬇️ Fetching page ${page} of address confirmed transactions: `, this.hash)
-
-    const { data } = await client.explorer.getAddressTransactions(this.hash, page)
-
-    const isInitialData = page === 1 && data.length > 0 && this.transactions.confirmed.length === 0
-    const latestTxHashIsNew =
-      page === 1 &&
-      data.length > 0 &&
-      this.transactions.confirmed.length > 0 &&
-      data[0].hash !== this.transactions.confirmed[0].hash
-
-    if (isInitialData) {
-      this.transactions.confirmed = data
-      this.transactions.loadedPage = 1
-    } else if (latestTxHashIsNew || page > 1) {
-      const newTransactions = data.filter(
-        (tx) => !this.transactions.confirmed.find((confirmedTx) => confirmedTx.hash === tx.hash)
-      )
-      this.transactions.confirmed =
-        page > 1
-          ? this.transactions.confirmed.concat(newTransactions)
-          : newTransactions.concat(this.transactions.confirmed)
-      if (page > 1) {
-        this.transactions.loadedPage = page
-      }
-    }
-
-    this.lastUsed = this.transactions.confirmed.length > 0 ? this.transactions.confirmed[0].timestamp : this.lastUsed
-
-    return data
-  }
-
-  async fetchConfirmedTransactionsNextPage(client: Client) {
-    if (!client) throw new Error('No client provided')
-    await this.fetchConfirmedTransactions(client, this.transactions.loadedPage + 1)
-  }
-
   addPendingTransaction(transaction: SimpleTx) {
     console.log('🔵 Adding pending transaction sent from address: ', transaction.fromAddress)
 
@@ -168,47 +116,6 @@ export class Address {
       : this.transactions.pending.reduce((acc, tx) => (tx.amount ? acc + BigInt(tx.amount) : acc), BigInt(0))
     this.availableBalance = this.availableBalance - totalAmountOfPendingTxs
   }
-
-  async buildSweepTransactions(client: Client, toHash: AddressHash) {
-    if (!client) throw new Error('No client provided')
-
-    const { data } = await client.clique.transactionConsolidateUTXOs(this.publicKey, this.hash, toHash)
-    const fees = data.unsignedTxs.reduce((acc, tx) => acc + BigInt(tx.gasPrice) * BigInt(tx.gasAmount), BigInt(0))
-
-    return {
-      unsignedTxs: data.unsignedTxs,
-      fees
-    }
-  }
-
-  async signAndSendTransaction(
-    client: Client,
-    txId: string,
-    unsignedTx: string,
-    toHash: AddressHash,
-    type: TransactionType,
-    network: NetworkType,
-    amount?: bigint
-  ) {
-    if (!client) throw new Error('No client provided')
-
-    const signature = client.clique.transactionSign(txId, this.privateKey)
-    const response = await client.clique.transactionSend(toHash, unsignedTx, signature)
-
-    if (response.data) {
-      this.addPendingTransaction({
-        txId: response.data.txId,
-        fromAddress: this.hash,
-        toAddress: toHash,
-        timestamp: new Date().getTime(),
-        amount,
-        type,
-        network
-      })
-    }
-
-    return response.data
-  }
 }
 
 export type AddressesStateMap = Map<AddressHash, Address>
@@ -311,8 +218,8 @@ export const AddressesContextProvider: FC<{ overrideContextValue?: PartialDeep<A
 
       for (const address of addressesStateToRefresh) {
         try {
-          await address.fetchDetails(client)
-          await address.fetchConfirmedTransactions(client)
+          await client.fetchAddressDetails(address)
+          await client.fetchAddressConfirmedTransactions(address)
 
           const initialNumberOfPendingTransactions = address.transactions.pending.length
 
@@ -348,7 +255,7 @@ export const AddressesContextProvider: FC<{ overrideContextValue?: PartialDeep<A
   const fetchAddressTransactionsNextPage = async (address: Address) => {
     if (!client) return
     setIsLoadingData(true)
-    await address.fetchConfirmedTransactionsNextPage(client)
+    await client.fetchAddressConfirmedTransactionsNextPage(address)
     setIsLoadingData(false)
   }
 
diff --git a/src/pages/Wallet/AddressSweepModal.tsx b/src/pages/Wallet/AddressSweepModal.tsx
index 2acf3e5..d72f602 100644
--- a/src/pages/Wallet/AddressSweepModal.tsx
+++ b/src/pages/Wallet/AddressSweepModal.tsx
@@ -60,7 +60,7 @@ const AddressSweepModal = ({ sweepAddress, onClose, onSuccessfulSweep }: Address
       if (!client || !sweepAddresses.from || !sweepAddresses.to) return
       setIsLoading(true)
       try {
-        const { unsignedTxs, fees } = await sweepAddresses.from.buildSweepTransactions(client, sweepAddresses.to.hash)
+        const { unsignedTxs, fees } = await client.buildSweepTransactions(sweepAddresses.from, sweepAddresses.to.hash)
         setBuiltUnsignedTxs(unsignedTxs)
         setFee(fees)
       } catch (e) {
@@ -81,8 +81,8 @@ const AddressSweepModal = ({ sweepAddress, onClose, onSuccessfulSweep }: Address
     setIsLoading(true)
     try {
       for (const { txId, unsignedTx } of builtUnsignedTxs) {
-        const txSendResp = await sweepAddresses.from.signAndSendTransaction(
-          client,
+        const txSendResp = await client.signAndSendTransaction(
+          sweepAddresses.from,
           txId,
           unsignedTx,
           sweepAddresses.from.hash,
diff --git a/src/pages/Wallet/SendModal.tsx b/src/pages/Wallet/SendModal.tsx
index 480f835..b15b732 100644
--- a/src/pages/Wallet/SendModal.tsx
+++ b/src/pages/Wallet/SendModal.tsx
@@ -103,7 +103,7 @@ const SendModal = ({ title, onClose }: SendModalProps) => {
 
       try {
         if (sweep) {
-          const { unsignedTxs, fees } = await fromAddress.buildSweepTransactions(client, toAddress)
+          const { unsignedTxs, fees } = await client.buildSweepTransactions(fromAddress, toAddress)
           setSweepUnsignedTxs(unsignedTxs)
           setFees(fees)
         } else {
@@ -153,7 +153,7 @@ const SendModal = ({ title, onClose }: SendModalProps) => {
 
       setIsLoading(true)
       const { fromAddress } = transactionData
-      const { unsignedTxs, fees } = await fromAddress.buildSweepTransactions(client, fromAddress.hash)
+      const { unsignedTxs, fees } = await client.buildSweepTransactions(fromAddress, fromAddress.hash)
       setSweepUnsignedTxs(unsignedTxs)
       setFees(fees)
       setIsLoading(false)
@@ -176,8 +176,8 @@ const SendModal = ({ title, onClose }: SendModalProps) => {
           const transactionType = consolidationRequired ? 'consolidation' : 'sweep'
 
           for (const { txId, unsignedTx } of sweepUnsignedTxs) {
-            const data = await fromAddress.signAndSendTransaction(
-              client,
+            const data = await client.signAndSendTransaction(
+              fromAddress,
               txId,
               unsignedTx,
               sendToAddress,
@@ -190,8 +190,8 @@ const SendModal = ({ title, onClose }: SendModalProps) => {
             }
           }
         } else {
-          const data = await fromAddress.signAndSendTransaction(
-            client,
+          const data = await client.signAndSendTransaction(
+            fromAddress,
             unsignedTxId,
             unsignedTransaction,
             toAddress,
diff --git a/src/utils/api-clients.ts b/src/utils/api-clients.ts
index 394666f..82793ec 100644
--- a/src/utils/api-clients.ts
+++ b/src/utils/api-clients.ts
@@ -18,7 +18,8 @@ along with the library. If not, see <http://www.gnu.org/licenses/>.
 
 import { CliqueClient, ExplorerClient } from 'alephium-js'
 
-import { loadStoredSettings, Settings } from './settings'
+import { Address, AddressHash, TransactionType } from '../contexts/addresses'
+import { loadStoredSettings, NetworkType, Settings } from './settings'
 
 export async function createClient(settings?: Settings['network']) {
   const loadedSettings = settings || loadStoredSettings().network
@@ -42,7 +43,102 @@ export async function createClient(settings?: Settings['network']) {
     // Init clients
     await cliqueClient.init(isMultiNodesClique)
 
-    return { clique: cliqueClient, explorer: explorerClient }
+    const fetchAddressDetails = async (address: Address) => {
+      console.log('⬇️ Fetching address details: ', address.hash)
+
+      const { data } = await explorerClient.getAddressDetails(address.hash)
+      address.details = data
+
+      if (data.balance) address.availableBalance = BigInt(data.balance)
+      if (data.lockedBalance) address.availableBalance -= BigInt(data.lockedBalance)
+
+      return data
+    }
+
+    const fetchAddressConfirmedTransactions = async (address: Address, page = 1) => {
+      console.log(`⬇️ Fetching page ${page} of address confirmed transactions: `, address.hash)
+
+      const { data } = await explorerClient.getAddressTransactions(address.hash, page)
+
+      const isInitialData = page === 1 && data.length > 0 && address.transactions.confirmed.length === 0
+      const latestTxHashIsNew =
+        page === 1 &&
+        data.length > 0 &&
+        address.transactions.confirmed.length > 0 &&
+        data[0].hash !== address.transactions.confirmed[0].hash
+
+      if (isInitialData) {
+        address.transactions.confirmed = data
+        address.transactions.loadedPage = 1
+      } else if (latestTxHashIsNew || page > 1) {
+        const newTransactions = data.filter(
+          (tx) => !address.transactions.confirmed.find((confirmedTx) => confirmedTx.hash === tx.hash)
+        )
+        address.transactions.confirmed =
+          page > 1
+            ? address.transactions.confirmed.concat(newTransactions)
+            : newTransactions.concat(address.transactions.confirmed)
+        if (page > 1) {
+          address.transactions.loadedPage = page
+        }
+      }
+
+      address.lastUsed =
+        address.transactions.confirmed.length > 0 ? address.transactions.confirmed[0].timestamp : address.lastUsed
+
+      return data
+    }
+
+    const fetchAddressConfirmedTransactionsNextPage = async (address: Address) => {
+      await fetchAddressConfirmedTransactions(address, address.transactions.loadedPage + 1)
+    }
+
+    const buildSweepTransactions = async (address: Address, toHash: AddressHash) => {
+      const { data } = await cliqueClient.transactionConsolidateUTXOs(address.publicKey, address.hash, toHash)
+      const fees = data.unsignedTxs.reduce((acc, tx) => acc + BigInt(tx.gasPrice) * BigInt(tx.gasAmount), BigInt(0))
+
+      return {
+        unsignedTxs: data.unsignedTxs,
+        fees
+      }
+    }
+
+    const signAndSendTransaction = async (
+      address: Address,
+      txId: string,
+      unsignedTx: string,
+      toHash: AddressHash,
+      type: TransactionType,
+      network: NetworkType,
+      amount?: bigint
+    ) => {
+      const signature = cliqueClient.transactionSign(txId, address.privateKey)
+      const response = await cliqueClient.transactionSend(toHash, unsignedTx, signature)
+
+      if (response.data) {
+        address.addPendingTransaction({
+          txId: response.data.txId,
+          fromAddress: address.hash,
+          toAddress: toHash,
+          timestamp: new Date().getTime(),
+          amount,
+          type,
+          network
+        })
+      }
+
+      return response.data
+    }
+
+    return {
+      clique: cliqueClient,
+      explorer: explorerClient,
+      fetchAddressDetails,
+      fetchAddressConfirmedTransactions,
+      fetchAddressConfirmedTransactionsNextPage,
+      buildSweepTransactions,
+      signAndSendTransaction
+    }
   } catch (error) {
     console.error(error)
     return false
```
</details>

Reply:
Yes! It makes way more sense to me. 💯